traversal/TraversalEngine.java

```diff
@@ -47,21 +47,19 @@ public GraphManager graph() {
 
     public FunctionalProducer<VertexMap> producer(GraphTraversal.Thing traversal, Either<Arguments.Query.Producer, Long> context,
                                                   int parallelisation) {
-        traversal.initialise(cache);
-        return traversal.producer(graphMgr, context, parallelisation);
+        return traversal.permutationProducerCaching(graphMgr, cache, context, parallelisation);
     }
 
     public FunctionalIterator<VertexMap> iterator(GraphTraversal.Thing traversal) {
-        traversal.initialise(cache);
-        return traversal.permutation(graphMgr);
```

Comment:
we allow the traversal to manage the cache directly, rather than calling `initialise()` from outside